flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -363,1691 +372,1780 @@ public int getKeyGroupPrefixBytes() {
-					kvStateIterators = null;
+				// Check that restored key groups all belong to the backend
+				Preconditions.checkState(rocksDBKeyedStateBackend.getKeyGroupRange().contains(keyGroup),
+					"The key group must belong to the backend");
 
-					//preamble: setup with first key-group as our lookahead
-					if (mergeIterator.isValid()) {
-						//begin first key-group by recording the offset
-						keyGroupRangeOffsets.setKeyGroupOffset(mergeIterator.keyGroup(), outStream.getPos());
-						//write the k/v-state id as metadata
-						kgOutStream = stateBackend.keyGroupCompressionDecorator.decorateWithCompression(outStream);
-						kgOutView = new DataOutputViewStreamWrapper(kgOutStream);
+				long offset = keyGroupOffset.f1;
+				//not empty key-group?
+				if (0L != offset) {
+					currentStateHandleInStream.seek(offset);
+					try (InputStream compressedKgIn = keygroupStreamCompressionDecorator.decorateWithCompression(currentStateHandleInStream)) {
+						DataInputViewStreamWrapper compressedKgInputView = new DataInputViewStreamWrapper(compressedKgIn);
 						//TODO this could be aware of keyGroupPrefixBytes and write only one byte if possible
-						kgOutView.writeShort(mergeIterator.kvStateId());
-						previousKey = mergeIterator.key();
-						previousValue = mergeIterator.value();
-						mergeIterator.next();
+						int kvStateId = compressedKgInputView.readShort();
+						ColumnFamilyHandle handle = currentStateHandleKVStateColumnFamilies.get(kvStateId);
+						//insert all k/v pairs into DB
+						boolean keyGroupHasMoreKeys = true;
+						while (keyGroupHasMoreKeys) {
+							byte[] key = BytePrimitiveArraySerializer.INSTANCE.deserialize(compressedKgInputView);
+							byte[] value = BytePrimitiveArraySerializer.INSTANCE.deserialize(compressedKgInputView);
+							if (RocksDBFullSnapshotOperation.hasMetaDataFollowsFlag(key)) {
+								//clear the signal bit in the key to make it ready for insertion again
+								RocksDBFullSnapshotOperation.clearMetaDataFollowsFlag(key);
+								rocksDBKeyedStateBackend.db.put(handle, key, value);
+								//TODO this could be aware of keyGroupPrefixBytes and write only one byte if possible
+								kvStateId = RocksDBFullSnapshotOperation.END_OF_KEY_GROUP_MARK
+									& compressedKgInputView.readShort();
+								if (RocksDBFullSnapshotOperation.END_OF_KEY_GROUP_MARK == kvStateId) {
+									keyGroupHasMoreKeys = false;
+								} else {
+									handle = currentStateHandleKVStateColumnFamilies.get(kvStateId);
+								}
+							} else {
+								rocksDBKeyedStateBackend.db.put(handle, key, value);
+							}
+						}
 					}
+				}
+			}
+		}
+	}
 
-					//main loop: write k/v pairs ordered by (key-group, kv-state), thereby tracking key-group offsets.
-					while (mergeIterator.isValid()) {
+	/**
+	 * Encapsulates the process of restoring a RocksDBKeyedStateBackend from an incremental snapshot.
+	 */
+	private static class RocksDBIncrementalRestoreOperation<T> {
```

Comment:
Would it actually make sense to give these classes their own files? The `RocksDBKeyedStateBackend` file is 2100 lines long which makes it quite hard to handle.

Reply:
In general, I think that makes sense by now, but I would do this in a different PR.